codesign/lib/src/file_codesign_visitor.dart

```diff
@@ -124,6 +125,66 @@ update these file paths accordingly.
     await tempDir.delete(recursive: true);
   }
 
+  /// Retrieve eninge artifact from google cloud storage and kick start a
+  /// recursive visit of its contents.
+  ///
+  /// Invokes [visitDirectory] to recursively visit the contents of the remote
+  /// zip. Also downloads, notarizes and uploads the engine artifact.
+  Future<void> processRemoteZip({
+    required String artifactFilePath,
+    required Directory parentDirectory,
+  }) async {
+    final FileSystem fs = tempDir.fileSystem;
+
+    // namespace by hashcode otherwise there will be collisions
+    final String localFilePath = '${artifactFilePath.hashCode}_${fs.path.basename(artifactFilePath)}';
+
+    // download the zip file
+    final File originalFile = await download(
+      remotePath: artifactFilePath,
+      localPath: remoteDownloadsDir.childFile(localFilePath).path,
+      commitHash: commitHash,
+      processManager: processManager,
+      tempDir: tempDir,
+    );
+
+    await unzip(
+      inputZip: originalFile,
+      outDir: parentDirectory,
+      processManager: processManager,
+    );
+
+    //extract entitlements file.
+    fileWithEntitlements = await parseEntitlements(parentDirectory, true);
+    fileWithoutEntitlements = await parseEntitlements(parentDirectory, false);
+    log.info('parsed binaries with entitlements are $fileWithEntitlements');
+    log.info('parsed binaries without entitlements are $fileWithEntitlements');
+
+    // recursively visit extracted files
+    await visitDirectory(directory: parentDirectory, entitlementParentPath: artifactFilePath);
+
+    final File codesignedFile = codesignedZipsDir.childFile(localFilePath);
+
+    await zip(
+      inputDir: parentDirectory,
+      outputZip: codesignedFile,
+      processManager: processManager,
+    );
+
+    // notarize
+    await notarize(codesignedFile);
+
+    // we will only upload for production
+    if (production) {
```

Comment:
The production flag is an antipattern. See other comment about having a Gsutil class that can be injected

Reply:
Sure, changed to an injected class. the class will do no-op if not in production mode.

Reply:
Seems like the class wasn't uploaded.

Reply:
umm I invoked the upload function with the new class like
```    
await engineArtifactTransfer.uploadEngineArtifact(
      localPath: codesignedFile.path,
      remotePath: artifactFilePath,
      commitHash: commitHash,
      processManager: processManager,
    );
```
umm I think the engine artifact will be uploaded. umm would be great if Casey could help elaborate a little bit on the class wasn't uploaded part

Reply:
It would be easier to find if it was in it's own file, and not hidden in utils.dart.

Also, the global functions in utils.dart are still concerning.

Reply:
Sure, I refactored it into its own file transfer_utils.dart.

And sure, I shifted the two global functions (upload and download) into within the EngineArtifactTransfer class, and provided default implementations of them within the class. The class will use default implementations of upload/download if nothing is passed in, and will use user defined download/upload functions if they are passed in.